aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java

```diff
@@ -109,7 +112,29 @@ private ThrowableDescription describeThrowable(Throwable throwable, String id) {
 
     @Override
     public List<ThrowableDescription> describeInContext(Throwable throwable, List<Subsegment> subsegments) {
+        return describeInContext(null, throwable, subsegments);
+    }
+
+    @Override
+    public List<ThrowableDescription> describeInContext(
+        @Nullable Entity entity,
+        Throwable throwable,
+        List<Subsegment> subsegments
+    ) {
         List<ThrowableDescription> result = new ArrayList<>();
+        IdGenerator idGenerator;
+
+        // We could fall back on ThreadLocalStorage#get if we were confident that no `SegmentContext` implementations
+        // override the default get/setTraceEntity implementations, but we can't make that guarantee.
+        if (entity != null) {
+            idGenerator = entity.getCreator().getIdGenerator();
+        } else if (!subsegments.isEmpty()) {
+            idGenerator = subsegments.get(0).getCreator().getIdGenerator();
+        } else {
+            // to maintain backwards compatibility, always fall back to the secure ID generator when we can't
+            // find a different one set
+            idGenerator = SecureIdGenerator.INSTANCE;
```

Comment:
how would you feel about me using the ID generator from the global recorder (via `AWSXRay::getGlobalRecorder`) instead of this?

Reply:
and same for in `TraceID::fromString`/`TraceID::create`

Reply:
Ah I'd actually prefer that! Since that will also default to secure if the user hasn't configured their recorder, but at least has a chance of picking up the configured value if the first two conditions don't get hit somehow.

Reply:
Are users allowed to set the global recorder to `null`? Most code assumes it won't return null, but there's no explicit check in `setGlobalRecorder`. Further, there's a test that even sets it to null, alongside some code in `TracingHandler` that's meant to deal with `getGlobalRecorder` returning null (although this code also assumes that all subsequent invocations of `getGlobalRecorder` will return an object).

Reply:
We added `Nullable` annotations throughout the SDK and left the `globalRecorder` field and corresponding getter in `AWSXRay` as not `Nullable` so I think it's safe to assume that customers are respecting the non-null contract. In `TracingHandler` I think we're protecting against the `TracingHandler` being constructed with a `null` recorder parameter, and then assume that `getGlobalRecorder` will provide the recorder if so. What test sets it to null?

Reply:
`testRaceConditionOnRecorderInitialization` in `TracingHandlerTest`

Reply:
Ah yeah that's unrelated to testing if the customer has set the global recorder to null, it's [this issue](https://github.com/aws/aws-xray-sdk-java/issues/29) where the AWS SDK's auto-instrumentor would create a `TracingHandler` and call `AWSXRay.getGlobalRecorder` in parallel with the recorder's start up before it had finished initializing. Not something that we'd expect in regular customer code that's run after the recorder is initialized.